cloudinit/config/schema.py

```diff
@@ -66,23 +65,25 @@
 DEPRECATED_KEY = "deprecated"
 
 
-# annotations add value for development, but don't break old versions
-# pyver: 3.6 -> 3.8
-# pylint: disable=E1101
-if sys.version_info >= (3, 8):
+# type-annotate only if type-checking.
```

Comment:
If we do not define `skip_by_schema` as `NotRequired`, then mypy is going to raise type-exceptions for `MetaSchema` instances without the new property defined (I am thinking about downstream custom modules).

For example, if we leave this as it was but with the new property:

```py
# annotations add value for development, but don't break old versions
# pyver: 3.6 -> 3.8
# pylint: disable=E1101
if sys.version_info >= (3, 8):

    class MetaSchema(typing.TypedDict):
        name: str
        id: str
        title: str
        description: str
        distros: typing.List[str]
        examples: typing.List[str]
        frequency: str
        skip_by_schema: typing.List[str]

else:
    MetaSchema = dict
# pylint: enable=E1101
```

and we run mypy with a python version >= 3.8, then we get:

```sh
$ tox -e mypy
...
cloudinit/config/cc_byobu.py:35: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_bootcmd.py:23: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_apt_pipelining.py:27: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_apt_configure.py:35: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_apk_configure.py:56: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
tests/unittests/config/test_schema.py:715: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
tests/unittests/config/test_modules.py:42: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
tests/unittests/config/test_modules.py:75: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
cloudinit/config/cc_write_files_deferred.py:23: error: Missing key "skip_by_schema" for TypedDict "MetaSchema"
Found 62 errors in 61 files (checked 489 source files)
ERROR: InvocationError for command /cloud-init/.tox/mypy/bin/python -m mypy cloudinit/ tests/ tools/ (exited with code 1)
_____________________________________________________________________________________________________ summary _____________________________________________________________________________________________________
ERROR:   mypy: commands failed

```
Similarly, if we do not include `skip_by_schema: typing.List[str]` at all, then we get error from schemas with the property.

This is not a problem for us, as we could add the key to all our meta schemas. But could hit downstream projects with custom modules.

If we do not want to make use of `type_extensions`, then we could tell downstream projects:

1. Add `"skip_by_schema": []` to their custom modules.
2. Or remove this if block and define `MetaSchema = dict` as a temporal solution, but apply 1 at some point.

More context on: https://github.com/canonical/cloud-init/pull/1589

Reply:
After a team discussion, we have agreed on accepting the risk and use `type_extensions` only at type-checking time in this case.